packages/hyperliquid-composer/contracts/extensions/PreFundedFeeAbstraction.sol

```diff
@@ -0,0 +1,229 @@
+        if (assetIndex != ERC20_CORE_INDEX_ID) revert InvalidSpot();
+        SPOT_PAIR_ID = _spotPairId;
+
+        TokenInfo memory baseAssetInfo = _tokenInfo(assetIndex);
+        TokenInfo memory quoteAssetInfo = _tokenInfo(QUOTE_ASSET_INDEX);
+
+        /// @dev Hyperliquid protocol uses (8 - szDecimals) for spot price decimal encoding
+        SPOT_PRICE_DECIMALS = uint64(10 ** (SPOT_PRICE_MAX_DECIMALS - baseAssetInfo.szDecimals));
+        QUOTE_ASSET_DECIMALS = uint64(10 ** quoteAssetInfo.weiDecimals);
+
+        uint64 totalCentsAmount = BASE_ACTIVATION_FEE_CENTS + _activationOverheadFee;
+        ACTIVATION_COST = uint64((totalCentsAmount * QUOTE_ASSET_DECIMALS) / 100);
+
+        if (MIN_USD_PRE_FUND_AMOUNT() * QUOTE_ASSET_DECIMALS > type(uint64).max) revert MinUSDAmtGreaterThanU64Max();
+    }
+
+    /**
+     * @notice Override to track activation fees during transfers
+     * @param _to Destination address on HyperCore
+     * @param _amountLD Amount to transfer in LayerZero decimals
+     */
+    function _transferERC20ToHyperCore(address _to, uint256 _amountLD) internal virtual override {
+        IHyperAssetAmount memory amounts = quoteHyperCoreAmount(
+            ERC20_CORE_INDEX_ID,
+            ERC20_DECIMAL_DIFF,
+            ERC20_ASSET_BRIDGE,
+            _amountLD
+        );
+
+        if (amounts.evm != 0) {
+            uint64 originalAmount = amounts.core;
+            uint64 coreAmount = _getFinalCoreAmount(_to, originalAmount);
+
+            /// @dev When the user is not activated we collect the activation fee
+            if (originalAmount > coreAmount) {
+                uint64 coreBalance = spotBalance(address(this), QUOTE_ASSET_INDEX).total;
+                if (coreBalance < MIN_USD_PRE_FUND_WEI_VALUE()) revert InsufficientCoreAmountForActivation();
+
+                uint64 feeCollected = originalAmount - coreAmount;
+                accruedActivationFees += feeCollected;
+                emit FeeCollected(feeCollected);
+            }
+
+            IERC20(ERC20).safeTransfer(ERC20_ASSET_BRIDGE, amounts.evm);
+            _submitCoreWriterTransfer(_to, ERC20_CORE_INDEX_ID, coreAmount);
+        }
+    }
+
+    /**
+     * @notice Calculates activation fee in asset tokens using current spot price
+     * @return Activation fee amount in core asset decimals
+     */
+    function activationFee() public view virtual override returns (uint64) {
+        uint64 rawPrice = _spotPx(SPOT_PAIR_ID);
+        return uint64((ACTIVATION_COST * SPOT_PRICE_DECIMALS) / rawPrice);
+    }
+
+    /**
+     * @notice Estimates USD value of accumulated fees using current spot price
+     * @return USD value in quote token terms
+     */
+    function getAccruedFeeUsdValue() public view returns (uint256) {
+        if (accruedActivationFees == 0) return 0;
+
+        uint64 rawPrice = _spotPx(SPOT_PAIR_ID);
+        return (rawPrice * accruedActivationFees) / SPOT_PRICE_DECIMALS;
+    }
```

Comment:
You know what would be better than having this function?

Not having this function.
```suggestion
```